transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/utils/TrinoKeywordsConverter.java

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright 2019 LinkedIn Corporation. All rights reserved.
+ * Licensed under the BSD-2 Clause license.
+ * See LICENSE in the project root for license information.
+ */
+package com.linkedin.transport.trino.utils;
+
+import com.google.common.collect.ImmutableSet;
+import java.util.Locale;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+
+/**
+ * Trino Keywords Converter maintains the Trino Keywords and related utility functions
+ */
+public final class TrinoKeywordsConverter {
+
+  private TrinoKeywordsConverter() {
+
+  }
+
+  private static final Set<String> RESERVED_KEYWORDS =
+      ImmutableSet.of("ALTER", "AND", "AS", "BETWEEN", "BY", "CASE", "CAST", "CONSTRAINT", "CREATE", "CROSS", "CUBE",
+          "CURRENT_DATE", "CURRENT_TIME", "CURRENT_TIMESTAMP", "CURRENT_USER", "DEALLOCATE", "DELETE", "DESCRIBE",
+          "DISTINCT", "DROP", "ELSE", "END", "ESCAPE", "EXCEPT", "EXECUTE", "EXISTS", "EXTRACT", "FALSE", "FOR", "FROM",
+          "FULL", "GROUP", "GROUPING", "HAVING", "IN", "INNER", "INSERT", "INTERSECT", "INTO", "IS", "JOIN", "LEFT",
+          "LIKE", "LOCALTIME", "LOCALTIMESTAMP", "NATURAL", "NORMALIZE", "NOT", "NULL", "ON", "OR", "ORDER", "OUTER",
+          "PREPARE", "RECURSIVE", "RIGHT", "ROLLUP", "SELECT", "TABLE", "THEN", "TRUE", "UESCAPE", "UNION", "UNNEST",
+          "USING", "VALUES", "WHEN", "WHERE", "WITH");
+
+  /**
+   * Quote the preserved keywords which might appear as field names in the type signatures
+   * i.e. `row(key varchar, values varchar)` will be converted to `row(key varchar, "values" varchar)`
+   * if they are not quoted, `io.trino.sql.analyzer.TypeSignatureTranslator.parseTypeSignature` will throw exception
+   *
+   * @param signature input type signature
+   * @return converted type signature which is properly quoted
+   */
+  public static String quoteReservedKeywordsInTypeSignature(String signature) {
+    signature = signature.toLowerCase(Locale.ROOT);
+    for (String keyword : RESERVED_KEYWORDS) {
+      String lowercaseKeyword = keyword.toLowerCase(Locale.ROOT);
+      // the preserved keyword may only appear as a field name in `row` type
+      // in the following scenarios (`values` is the preserved keyword):
+      // (1) row(field1 type1, values type2)
+      // (2) row(field1 type1,values type2)
+      // (3) row(values type1, field2 type2)
+      // therefore, the previous character must be one of `,`, `(` or space(s)
+      // and there must be the type name following the keyword after space(s)
+      Pattern pattern = Pattern.compile("(\\(|\\s+|,)" + lowercaseKeyword + "\\s+\\w+");
+      final Matcher matcher = pattern.matcher(signature);
+      while (matcher.find()) {
+        final String group = matcher.group();
+        final String replacedGroup = group.replaceAll(lowercaseKeyword, "\"" + lowercaseKeyword + "\"");
```

Comment:
Should this be `replaceFirst` to avoid replacing the second `VALUES` in `row(VALUES VALUES)`? The second `VALUES` is a generic type parameter. Generally I suggest parsing the type signature, replacing, and converting back to string since regex approach can be a bit fragile. Check `parse` and `toString` functions in `TypeSignature` class.

Reply:
Agreed with `replaceFirst`.

I tried to use `TypeSignature`, but it's not suitable for our use case.
IIUC, your suggestion is 
```
1. parse signature string to get typeSignature
2. Traverse parameters of typeSignature, construct newTypeSignature with modified parameters and quoted parameterNames
3. newTypeSignature.toString 
```
there are several issues:
In step 2, there is no public constructor to construct typeSignature with parameters & parameterNames, we need to expose a protected constructor for that. Besides, it will throw exception [here](https://github.com/linkedin/transport/blob/master/transportable-udfs-type-system/src/main/java/com/linkedin/transport/typesystem/TypeSignature.java#L188) if the field name is quoted during the construction.
In step 3, `typeSignature.toString` will change the type name. (i.e. `varchar -> string`, `row -> struct`), not aligned with the Trino parser's expectation.

So I still prefer to keep using current regex if we can't come up with a corner case.

Reply:
Can we do our own `toTrinoTypeSignatureString`?

Reply:
Just use (1) from the above, and `toTrinoTypeSignatureString` handles the rest.

Reply:
What about the first issue I mentioned in step 2? Modify the existing `TypeSignature` or create a new one?

Reply:
I guess we do not have to create a new Type signature. We can traverse the original one, and output the quotes as part of the string construction.

Reply:
Thanks for clarifying, updated it.